### PR TITLE
Feature/solid

### PR DESCRIPTION
**Pull request by DmitryTumanov**
Some of code refactor using solid principles

### CODE REVIEW

File: OnlinerTask/OnlinerTask.Data/Repository/Implementations/MsSQLProductRepository.cs [hunk 1]
```diff
@@ -25,32 +22,63 @@ public MsSqlProductRepository(IRepository repository)
...
 
-        public async Task<bool> CheckItem(int itemId, string username)
+        public UsersUpdateEmail WriteUpdate(ProductModel item, string useremail)
         {
-            using (var context = new OnlinerProducts())
+            using (var db = new ApplicationDbContext())
             {
-                return await context.Product.FirstOrDefaultAsync(x => x.ProductId == itemId && x.UserEmail == username) != null;
+                var time = db.Users.FirstOrDefault(x => x.Email == useremail);
+                if (time != null)
```
Comment (andrei-piatrou) [inline OnlinerTask/OnlinerTask.Data/Repository/Implementations/MsSQLProductRepository.cs]: How about invert this if?

File: OnlinerTask/OnlinerTask.Data/Repository/Implementations/MsSQLProductRepository.cs [hunk 2]
```diff
@@ -25,32 +22,63 @@ public MsSqlProductRepository(IRepository repository)
...
 
-        public bool CreateOnlinerProduct(ProductModel model, string userEmail)
+        private UsersUpdateEmail WriteUpdateToProduct(ProductModel model, TimeSpan time)
         {
-            return repository.CreateOnlinerProduct(model, userEmail);
+            if (model == null)
+            {
+                return null;
+            }
+            using (var db = new OnlinerProducts())
```
Comment (andrei-piatrou) [inline OnlinerTask/OnlinerTask.Data/Repository/Implementations/MsSQLProductRepository.cs]: Dont like new statement for implementations. Lets create facroty for contexts or simple injection.

File: OnlinerTask/OnlinerTask.Data/Repository/Implementations/MsSQLRepository.cs [hunk 3]
```diff
@@ -57,34 +57,28 @@ private bool CreateProduct(Product product)
...
+            return true;
         }
 
         public async Task<string> RemoveOnlinerProduct(int itemId, string name)
         {
-            using (var context = new OnlinerProducts())
+            var model = await context.Product.FirstOrDefaultAsync(x => x.UserEmail == name && x.ProductId == itemId);
+            if (model == null)
+            {
+                return "";
```
Comment (andrei-piatrou) [inline OnlinerTask/OnlinerTask.Data/Repository/Implementations/MsSQLRepository.cs]: string.Empty?

**Review by andrei-piatrou — CHANGES_REQUESTED** [review]

File: OnlinerTask/OnlinerTask.BLL/Services/ConfigChange/IConfigChanger.cs [hunk 4]
```diff
@@ -0,0 +1,7 @@
+﻿namespace OnlinerTask.BLL.Services.ConfigChange
+{
+    public interface IConfigChanger
+    {
+        void TechnologySwap(string newValue);
```
Comment (andrei-piatrou) [inline OnlinerTask/OnlinerTask.BLL/Services/ConfigChange/IConfigChanger.cs]: Not clear name.

File: OnlinerTask/OnlinerTask.BLL/Services/ConfigChange/IConfigChanger.cs [hunk 5]
```diff
@@ -0,0 +1,7 @@
+﻿namespace OnlinerTask.BLL.Services.ConfigChange
+{
+    public interface IConfigChanger
```
Comment (andrei-piatrou) [inline OnlinerTask/OnlinerTask.BLL/Services/ConfigChange/IConfigChanger.cs]: Name is not clear.

**Review by andrei-piatrou — CHANGES_REQUESTED** [review]

File: OnlinerTask/OnlinerTask.BLL/Services/Search/ProductParser/IProductParser.cs [hunk 6]
```diff
@@ -0,0 +1,10 @@
+﻿using System.Net;
+using OnlinerTask.Data.SearchModels;
+
+namespace OnlinerTask.BLL.Services.Search.ProductParser
+{
+    public interface IProductParser
+    {
+        SearchResult ParseProductsFromRequest(HttpWebResponse webResponse);
```
Comment (andrei-piatrou) [inline OnlinerTask/OnlinerTask.BLL/Services/Search/ProductParser/IProductParser.cs]: FromRequest is easier name. Class name already contains part of it..

File: OnlinerTask/OnlinerTask.BLL/Services/Search/Request/Implementations/OnlinerRequestCreator.cs [hunk 7]
```diff
@@ -0,0 +1,16 @@
+﻿using System.Net;
+using OnlinerTask.Data.Resources;
+
+namespace OnlinerTask.BLL.Services.Search.Request.Implementations
+{
+    public class OnlinerRequestCreator : IRequestCreator
```
Comment (andrei-piatrou) [inline OnlinerTask/OnlinerTask.BLL/Services/Search/Request/Implementations/OnlinerRequestCreator.cs]: Facroty is common name for this type of classes.

**Review by andrei-piatrou — CHANGES_REQUESTED** [review]

**Review by andrei-piatrou — APPROVED** [review]

**Review by andrei-piatrou — APPROVED** [review]

File: OnlinerTask/OnlinerTask.Data/Notifications/Technologies/Implementations/NetMqNotificator.cs [hunk 8]
```diff
@@ -1,29 +1,39 @@
...
         }
 
         public void ChangeInfo(dynamic time)
         {
-            socket.ChangeInfo(time);
+            SendTcpMessage(time, Configurations.TcpConnectionPath, Configurations.InfoChatType);
         }
 
-        public void RemoveProduct(dynamic name, dynamic path = null)
+        private static void SendTcpMessage(string text, string tcpString, string type, string path = null)
```
Comment (andrei-piatrou) [inline OnlinerTask/OnlinerTask.Data/Notifications/Technologies/Implementations/NetMqNotificator.cs:23]: This method does a lot.

**Review by andrei-piatrou — CHANGES_REQUESTED** [review]

File: OnlinerTask/OnlinerTask.Data/Notifications/Technologies/Implementations/NetMqNotificator.cs [hunk 9]
```diff
@@ -25,15 +25,21 @@ private static void SendTcpMessage(string text, string tcpString, string type, s
             using (var client = NetMQContext.Create().CreateRequestSocket())
             {
                 client.Connect(tcpString);
```
Comment (andrei-piatrou) [inline OnlinerTask/OnlinerTask.Data/Notifications/Technologies/Implementations/NetMqNotificator.cs]: 27 and 28 lines can be swapped, can you answer why?

**Review by andrei-piatrou — APPROVED** [review]

File: OnlinerTask/OnlinerTask.Data/Notifications/Technologies/Implementations/NetMqNotificator.cs (hunk 9 above)
Comment (DmitryTumanov) [reply OnlinerTask/OnlinerTask.Data/Notifications/Technologies/Implementations/NetMqNotificator.cs]: Client doesn`t play any role in creating message, that`s why firstly we can create message and than configure client to sending

Comment (andrei-piatrou) [reply OnlinerTask/OnlinerTask.Data/Notifications/Technologies/Implementations/NetMqNotificator.cs]: Correct. so common logic should be grouped. Lets create message and then configure client and send this message.

**Review by andrei-piatrou — APPROVED** [review]

File: OnlinerTask/OnlinerTask.BLL/Services/Search/Implementations/SearchService.cs [hunk 10]
```diff
@@ -24,11 +24,11 @@ public SearchService(IProductRepository repository, IRequestFactory requestCreat
 
         public async Task<List<ProductModel>> GetProducts(SearchRequest searchRequest, string userName)
         {
-            if (string.IsNullOrEmpty(searchRequest?.SearchString))
+            if (string.IsNullOrEmpty(searchRequest?.SearchString) || searchRequest.PageNumber == 0)
```
Comment (andrei-piatrou) [inline OnlinerTask/OnlinerTask.BLL/Services/Search/Implementations/SearchService.cs]: possible < 1?

**Review by andrei-piatrou — CHANGES_REQUESTED** [review]

File: OnlinerTask/OnlinerTask.BLL/Services/Search/Implementations/SearchService.cs [hunk 11]
```diff
@@ -28,7 +29,7 @@ public SearchService(IProductRepository repository, IRequestFactory requestCreat
             {
                 return null;
             }
-            var request = requestCreator.CreateRequest(searchRequest.SearchString, searchRequest.PageNumber);
+            var request = requestCreator.CreateRequest(searchRequest.SearchString + Configurations.OnlinerPageVariable + searchRequest.PageNumber);
```
Comment (andrei-piatrou) [inline OnlinerTask/OnlinerTask.BLL/Services/Search/Implementations/SearchService.cs]: Factory?

File: OnlinerTask/OnlinerTask.BLL/Services/Search/Request/Implementations/OnlinerRequestFactory.cs [hunk 12]
```diff
@@ -5,10 +5,9 @@ namespace OnlinerTask.BLL.Services.Search.Request.Implementations
 {
     public class OnlinerRequestFactory : IRequestFactory
     {
-        public HttpWebRequest CreateRequest(string requestString, int page = 1)
+        public HttpWebRequest CreateRequest(string requestString)
```
Comment (andrei-piatrou) [inline OnlinerTask/OnlinerTask.BLL/Services/Search/Request/Implementations/OnlinerRequestFactory.cs]: Prefer building by parts (endpoint, params) or something.

**Review by andrei-piatrou — CHANGES_REQUESTED** [review]

**Review by andrei-piatrou — CHANGES_REQUESTED** [review]
Please create tests for this factory.

**Review by andrei-piatrou — CHANGES_REQUESTED** [review]
Please, add tests.

File: OnlinerTask/OnlinerTask.Tests/RequestFactoryTests.cs [hunk 13]
```diff
@@ -17,54 +15,101 @@ public IRequestFactory GetRequestFactory()
             return new OnlinerRequestFactory();
         }
 
-        public IDictionary<string, object> GetDictionary(IEnumerable<string> keys)
+        static readonly object[] TestDictionaries =
```
Comment (andrei-piatrou) [inline OnlinerTask/OnlinerTask.Tests/RequestFactoryTests.cs]: Access modifier is required.

**Review by andrei-piatrou — CHANGES_REQUESTED** [review]

**Review by andrei-piatrou — APPROVED** [review]

**Review by AndreiPiatrou — APPROVED** [review]
Not bad test cases.

**Review by andrei-piatrou — APPROVED** [review]

**Review by andrei-piatrou — APPROVED** [review]

**Review by andrei-piatrou — APPROVED** [review]

**Review by andrei-piatrou — APPROVED** [review]

**Review by andrei-piatrou — APPROVED** [review]

**Pull request merged**